src/devices/src/virtio/persist.rs

```diff
@@ -95,13 +109,122 @@ impl VirtioDeviceState {
     }
 }
 
+#[derive(Versionize)]
+// Clippy is complaining now because of the large size difference between variants.
+// Once we add Net state, the clippy exception can be removed.
+#[allow(clippy::large_enum_variant)]
+enum DeviceTypeState {
+    Block(BlockState),
+    // TODO: add net state when available
+    Net,
+    Vsock(VsockState),
+}
+
+#[derive(Versionize)]
+pub struct MmioTransportState {
+    // The register where feature bits are stored.
+    features_select: u32,
+    // The register where features page is selected.
+    acked_features_select: u32,
+    queue_select: u32,
+    device_status: u32,
+    config_generation: u32,
+    device_state: DeviceTypeState,
+}
+
+pub struct MmioTransportConstructorArgs {
+    mem: GuestMemoryMmap,
+}
+
+impl Persist for MmioTransport {
+    type State = MmioTransportState;
+    type ConstructorArgs = MmioTransportConstructorArgs;
+    type Error = Error;
+
+    fn save(&self) -> Self::State {
+        let virtio_dev = self.device();
+        let mut unlocked_dev = virtio_dev.lock().unwrap();
+        let device_state = match unlocked_dev.device_type() {
+            TYPE_BLOCK => {
```

Comment:
The code here should be agnostic to the device type. We must not add any FC specific dependencies here.  Please see my prev comments.

Reply:
Sounds good.. we can push the specific device type logic one (or more) levels up and keep the `MmioTransport` device-impl agnostic.

Reply:
done